### ISSUE TITLE
Support EKS Pod Identity

### ISSUE THREAD

**Issue by csantanapr** [argoproj/argo-cd#17064]
# Summary

EKS released in Nov/2023 [EKS Pod Identity](https://aws.amazon.com/blogs/aws/amazon-eks-pod-identity-simplifies-iam-permissions-for-applications-on-amazon-eks-clusters/), this is an alternative to [IAM roles for service accounts (IRSA)](https://aws.amazon.com/blogs/aws/amazon-eks-pod-identity-simplifies-iam-permissions-for-applications-on-amazon-eks-clusters/)

# Motivation

Have additional options on how ArgoCD authenticates with AWS other than using IRSA.

# Proposal

To support EKS Pod Identity the change is transparent if your using the aws-sdk, the version of the sdk needs to be updated to a recent version that supports EKS Pod Identity.

**Comment by csantanapr** [comment]
The PR that bumps the aws-sdk version is https://github.com/argoproj/argo-cd/pull/17063

A follow up PR will have the docs on how to configure ArgoCD to leverage the EKS Pod Identity

**Comment by todaywasawesome** [comment]
@csantanapr leaving this open until the docs are done

**Comment by csantanapr** [comment]
Sounds good @csantanapr I will be working on docs

**Comment by csantanapr** [comment]
@daniel-codefresh I was trying to work on docs but they were broken to build and preview, here is a fix https://github.com/argoproj/argo-cd/pull/17313

**Comment by tybook** [comment]
Hey @csantanapr do you have a branch with your in-progress docs? Or generally any rough guidance on how to configure ArgoCD to use EKS Pod Identity?

**Comment by tybook** [comment]
> > Hey @csantanapr do you have a branch with your in-progress docs? Or generally any rough guidance on how to configure ArgoCD to use EKS Pod Identity?
> 
> I'm not him obvi, but the solution is to just use a more recent release of argo-cd. The only necessary changes to support EKS Pod Identity are to have a more recent version of the AWS SDK running. If you are asking how to configure EKS Pod Identity in general, this guide has a bunch of information: https://docs.aws.amazon.com/eks/latest/userguide/pod-identities.html

I'm wondering specifically about how to configure an EKS cluster secret such that ArgoCD can authenticate with it via EKS Pod Identity instead of IRSA. The existing instructions at https://argo-cd.readthedocs.io/en/stable/operator-manual/declarative-setup/#eks assume the use of IRSA.

**Comment by AnubhavSabharwa** [comment]
Does argocd cluster secret support EKS Pod Identity

**Comment by magic-madrigal** [comment]
I've been trying to find a way to get it work for the past day and a half and haven't been successful. 

I feel I'm going to revert to IRSA, and then perhaps try using [eksctl's migration command](https://eksctl.io/usage/pod-identity-associations/#migrating-existing-iamserviceaccounts-and-addons-to-pod-identity-associations) to see what it suggests.

**Comment by AnubhavSabharwa** [comment]
@magic-madrigal We install argocd via another argocd as helm based application.
We are not allowed to save any token on the repository.
I have been trying this setup but at the end getting below error

**getting credentials: exec: executable argocd-k8s-auth failed with exit code 20**

**Comment by isikdos** [comment]
- Have an IAM role that is granted EKS Pod identity with an EKS Pod Identity Association https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/eks_pod_identity_association and don't forget the principal
- Have an IAM role that is given api access to the cluster that you want administrated by EKS Pod Identity. Make sure this can be assumed by the prior role.
- fill out the Argo auth brick for secret-type = cluster : https://argo-cd.readthedocs.io/en/stable/operator-manual/applicationset/Generators-Cluster/
- For the above, include the AWS Auth config from here https://argo-cd.readthedocs.io/en/stable/operator-manual/declarative-setup/
- Also make sure that you have the tlsclientConfig set properly

If you do all of that, it should work happily, I don't think I've forgotten any steps. You can always make sure of different components by looking at the pod running in EKS and seeing if it has the hallmark signs of using EKS pod identity, and also you can check the logs on the pod identity daemon to see what's going on

**Comment by magic-madrigal** [comment]
I appreciate the information @isikdos, but the application controller is still requesting additional credentials after going thought your write up. 

I'm quite certain that the roles are assuming each other, but they don't seem to be able to get the creds they need to deploy to the cluster. 

I've tried going the IRSA route, but yielded similar results. 

I just rebuilt my test clusters and I'm going to attempt a slightly different approach.

**Comment by AnubhavSabharwa** [comment]
@magic-madrigal I tried the IRSA thing, it is looking for additional credentials

**Comment by magic-madrigal** [comment]
@abhishekjiitr same here. All of the roles are assuming each other; I continue to get an error stating that the cluster is looking for credentials.

**Comment by magic-madrigal** [comment]
@abhishekjiitr - Progress!!! 

I updated to 2.12.0, but I don't think this was important to my success. 

I followed the IRSA instructions and was still having several issues with my implementation until I discovered an issue with my Cluster Secret. I had mistyped my clusterName in my secret, and once that was correct, it's now connecting correctly.

**Comment by mikestef9** [comment]
I have Pod Identity working with Argo CD, need to be using at least v2.10

```
  ArgoCDDeployerRole:
	Type: AWS::IAM::Role
	Properties:
	  RoleName: "ArgoCDDeployerRole"
	  Description: "Role for Argo CD deployment"
	  AssumeRolePolicyDocument:
		Version: '2012-10-17'
		Statement:
		  - Effect: Allow
			Principal:
			  AWS: "arn:aws:iam::123456789012:role/ArgoCDPodIdentityRole"
			Action:
			  - "sts:AssumeRole"
			  - "sts:TagSession"

  ArgoCDPodIdentityRole:
	Type: AWS::IAM::Role
	Properties:
	  RoleName: "ArgoCDPodIdentityRole"
	  Description: "Role to give Argo CD Access"
	  AssumeRolePolicyDocument:
		Version: '2012-10-17'
		Statement:
		  - Sid: "AllowEksAuthToAssumeRoleForPodIdentity"
			Effect: "Allow"
			Principal:
			  Service: "pods.eks.amazonaws.com"
			Action:
			  - "sts:AssumeRole"
			  - "sts:TagSession"
	  Policies:
		- PolicyName: AssumeRole
		  PolicyDocument:
			Version: "2012-10-17"
			Statement:
			  - Sid: "AssumeRole"
				Effect: "Allow"
				Action: "sts:AssumeRole"
				Resource: "arn:aws:iam::123456789012:role/ArgoCDDeployerRole"
				
```
Then in Helm chart config
```
configs:
  clusterCredentials:
    ipv4:
      server: "https://redacted.sk1.us-west-2.eks.amazonaws.com"
      labels:
        cluster-name: "ipv4"
      config:
        awsAuthConfig:
          clusterName: "ipv4"
          roleARN: "arn:aws:iam::123456789012:role/ArgoCDDeployerRole"
        tlsClientConfig:
          insecure: false
          caData: "redacted"
```

**Comment by AnubhavSabharwa** [comment]
@mikestef9 Thank you for sharing the things,
I have same config now but argocd is not able to add the cluster

**exec: executable argocd-k8s-auth failed with exit code 20**

**Comment by Utwo** [comment]
The instructions from @mikestef9 are good! On top of that, you need to add the access entries to the target cluster. https://argo-cd.readthedocs.io/en/stable/operator-manual/declarative-setup/#access-entries

I've managed to set it up with pod identity, multi-cluster, multi-account.

**Comment by BeyondEvil** [comment]
Mind sharing your configuration @Utwo ?

**Comment by AnubhavSabharwa** [comment]
Above configurations works.

Make sure you enable the acces in aws-auth.

**Comment by Utwo** [comment]
> Mind sharing your configuration @Utwo ?

I'm OOO this week. Will get back to you next week.

**Comment by scottt732** [comment]
I managed to get this working after a bit of trial & error based on @mikestef9's post. Hopefully this helps. Here is the portion of my terraform that deals with x-acct/pod identity:
https://gist.github.com/scottt732/fa53ba71164c7d8e936a87464ab9877d

I end up calling into this module 3 times, 1 from `env_code == "mgmt"` (the one where argo-cd gets installed) and 2 others (`env_code` == `dev` and `prod`). On these, I don't have ArgoCD installed.

- I created the `ArgoCDDeployerRole` role in all 3 accounts and granted the `ArgoCDPodIdentityRole` (which only exists in my mgmt account) the ability to assume it in all 3 accounts (its own included).
- `sts:TagSession` and `sts:AssumeRole` are both 100% required. 
- Make sure you have the pod identity addon installed on your cluster. It's there by default if you're using EKS Auto Mode (at least for the nodes managed by Auto mode)
- Make sure you restart argo-cd components after making service account/IAM-related changes. 
  ```
  kubectl -n argo-cd delete pod argo-cd-argocd-application-controller-0 && \
  kubectl rollout restart deployment/argo-cd-argocd-server -n argo-cd && \
  kubectl rollout restart deployment/argo-cd-argocd-applicationset-controller -n argo-cd && \
  kubectl rollout restart deployment/argo-cd-argocd-notifications-controller -n argo-cd && \
  kubectl rollout restart deployment/argo-cd-argocd-redis -n argo-cd && \
  kubectl rollout restart deployment/argo-cd-argocd-repo-server -n argo-cd
  ```
- Keep a close eye on `kubectl logs -n argo-cd -l app.kubernetes.io/name=argocd-application-controller --tail 30 --follow` after a restart. 
- I did not use the `aws-auth` config (it looks deprecated per argo-cd docs)
- If you see "Client disconnect before headers", the error message may be misleading. I was troubleshooting as a connectivity issue but it turned out to be related to the awsAuthConfig in my cluster secrets (I think)
- If you see "Server asked client to send credentials", make sure you [setup the Access Entries](https://argo-cd.readthedocs.io/en/stable/operator-manual/declarative-setup/#access-entries). In my case, I was much closer to a working solution once I started seeing this error.
- Spin one of these up in the same namespace as your argo-cd cluster to help troubleshoot:

```
apiVersion: v1
kind: Pod
metadata:
  name: aws-test
  namespace: argo-cd
spec:
  serviceAccountName: argocd-application-controller
  containers:
  - name: aws-cli
    image: amazon/aws-cli
    command: ["sleep", "1d"]
```

and see if you can get this working:
```
kubectl exec -it -n argo-cd aws-test -- aws sts assume-role \
  --role-arn "arn:aws:iam::${YOUR_AWS_ACCOUNT_GOES_HERE}:role/ArgoCDDeployerRole" \
  --role-session-name test-session
```

If you see something like this, you're close:
```
{
    "Credentials": {
        "AccessKeyId": "...",
        "SecretAccessKey": "...",
        "SessionToken": "...",
        "Expiration": "2025-02-11T19:49:30+00:00"
    },
    "AssumedRoleUser": {
        "AssumedRoleId": "...:test-session",
        "Arn": "arn:aws:sts::...:assumed-role/ArgoCDDeployerRole/test-session"
    },
    "PackedPolicySize": 66
}
```

**Comment by BeyondEvil** [comment]
Where does `caData` come from? The only mention I can find in docs is `<base64 encoded certificate>`.

What certificate? From where?

**Comment by BeyondEvil** [comment]
Thank you for that awesome writeup @scottt732

I've gotten to the point where the cluster is visible in the ArgoCD UI. But connection status is unknown and it says: 

> Cluster has no applications and is not being monitored.

Which is true. But I don't know how to check if it's truly working.

Note that I never restarted anything, and I don't see anything wonky in the logs.

**Comment by BeyondEvil** [comment]
OK, did the restart. Not seeing any warnings or errors, but I'm seeing:

```
time="2025-02-11T19:32:23Z" level=debug msg="Successfully saved info of 2 clusters"
```

So I guess everything is all good? 🤷

**Comment by scottt732** [comment]
I'm using `terraform-aws-modules/eks/aws` 20.31 ... and `caData` came from `module.eks[k].cluster_certificate_authority_data` in my setup. 

You can also pull it via `aws eks --profile <<YOUR AWS PROFILE>> describe-cluster --name <<YOUR CLUSTER NAME>> --output yaml` under `cluster.certificateAuthority.data`.

I'm not sure whether you need to have an app pointed at the cluster for it to try to pull cluster details. I was keeping an eye on the Settings -> Clusters page hoping to see it detect the k8s version (1.32 instead of Unknown or Failed). Not sure whether hitting the `Invalidate Caches` button on the details page was necessary/helpful.

**Comment by BeyondEvil** [comment]
Yeah, I pulled that caData from the AWS Console (we don't use terraform for helm).

Tried invalidating cache, but it didn't make a difference. I'll just continue adding an Application Set and take it from there.

Again, thank you for the great writeup! 🙏

**Comment by scottt732** [comment]
Ah yeah I forgot to mention that. It would be helpful if the clusters ui tried to grab the k8s version regardless of whether apps are setup to deploy there (yet) or not... or if there was an explicit refresh button on the details page. There was a lot of trial and error to get this working and it wasn't really clear whether Unknown meant "problem" or "no problem". It definitely wasn't giving the "problem solved" signal I was looking for.

**Comment by BeyondEvil** [comment]
> Ah yeah I forgot to mention that. It would be helpful if the clusters ui tried to grab the k8s version regardless of whether apps are setup to deploy there (yet) or not... or if there was an explicit refresh button on the details page. There was a lot of trial and error to get this working and it wasn't really clear whether Unknown meant "problem" or "no problem". It definitely wasn't giving the "problem solved" signal I was looking for.

This.

**Comment by starlightromero** [comment]
@scottt732 Thanks for this solution. I noticed you said
> the ArgoCDPodIdentityRole (which only exists in my mgmt account)

In the Gist, the resource `aws_iam_role.argocd_pod_identity_role`  does not have a count on the `var.env_code`. I assume this means it gets created on every module call. Is this correct?

**Comment by csantanapr** [comment]
It looks like there are a lot questions on how to setup eks pod identity using terraform for argocd. I will update the docs soon to update the section on IRSA to include Pod Identity and have some terraform examples in docs or links

**Comment by reggie-k** [comment]
TLDR: 
Argo CD already supports EKS Pod Identity, multiple configuration examples exist above. 
Official docs for this are currently being worked on.

**Comment by aminmiri** [comment]
+1

**Comment by Woitekku** [comment]
I was able to setup remote cluster connection with pod identity.

EKS: 1.33
ArgoCD: v3.1.8+becb020

Basically it's as-is mentioned in https://argo-cd.readthedocs.io/en/stable/operator-manual/declarative-setup/#eks
The ArgoCD role has to have "sts:TagSession" action allowed, next to "sts:AssumeRole" within its permissions.
The remote cluster role has to have "sts:TagSession" action allowed, next to "sts"AssumeRole" within its trusted relationship.

Everything else is the same, no OIDC/IRSA. Just make sure you do eks pod identity association.

**Comment by Woitekku** [comment]
@csantanapr check my PR. I can add some TF code snippets too.

**Comment by argais** [comment]
Heya, I was looking for these instructions :) amd I am a bit confused.

- [IRSA enabled](https://docs.aws.amazon.com/eks/latest/userguide/enable-iam-roles-for-service-accounts.html) or [Pod Identity agent](https://docs.aws.amazon.com/eks/latest/userguide/pod-id-agent-setup.html) on your Argo CD EKS cluster
- An IAM role ("management role") for your Argo CD EKS cluster that has an appropriate trust policy and permission policies (see below)
- A role created for each cluster being added to Argo CD that is assumable by the Argo CD management role
- An [Access Entry](https://docs.aws.amazon.com/eks/latest/userguide/access-entries.html) within each EKS cluster added to Argo CD that gives the cluster's role (from point 3) RBAC permissions to perform actions within the cluster

Why the role for each cluster if you can just add in each of them an access entry pointing to the management role in the argo cd eks cluster?

**Comment by BeyondEvil** [comment]
> Heya, I was looking for these instructions :) amd I am a bit confused.
> 
> * [IRSA enabled](https://docs.aws.amazon.com/eks/latest/userguide/enable-iam-roles-for-service-accounts.html) or [Pod Identity agent](https://docs.aws.amazon.com/eks/latest/userguide/pod-id-agent-setup.html) on your Argo CD EKS cluster
> * An IAM role ("management role") for your Argo CD EKS cluster that has an appropriate trust policy and permission policies (see below)
> * A role created for each cluster being added to Argo CD that is assumable by the Argo CD management role
> * An [Access Entry](https://docs.aws.amazon.com/eks/latest/userguide/access-entries.html) within each EKS cluster added to Argo CD that gives the cluster's role (from point 3) RBAC permissions to perform actions within the cluster
> 
> Why the role for each cluster if you can just add in each of them an access entry pointing to the management role in the argo cd eks cluster?

I guess the assumption here is that "each cluster" lives in their own accounts. But if the cluster(s) live in the same account as the cluster where ArgoCD is running, then they can all use the same role.

**Comment by moshiaiz** [comment]
We managed to get **EKS Pod Identity working with Argo CD cluster registration**, including **cross-account clusters**, so sharing what actually worked for us:

* We use **a single IAM role** in the **Argo CD (management) account**
  (no need to create a separate role per EKS cluster).
* That same role is granted access to **destination EKS clusters**, including clusters in **other AWS accounts**.
* On the **destination cluster** side:

  * Add the **Argo CD role ARN** as an **EKS Access Entry**
  * Attach an appropriate EKS access policy (for example `ClusterAdmin`)
* On the **Argo CD cluster**:

  * The **EKS Pod Identity Agent addon** must be installed
  * **Pod Identity associations** are created for:

    * `argocd-server`
    * `argocd-application-controller`
* We do **not** use `eks.amazonaws.com/role-arn` service account annotations at all

---

### Important

Since we use a single role, **`roleARN` must not be set in the cluster Secret**.

If `roleARN` is set, Argo CD will try to `AssumeRole` into the **same role it already has** via Pod Identity and will fail (unless provided with a weird self-assume trust policy).

Example cluster Secret (note the missing `roleARN`):

```yaml
metadata:
  name: destination-cluster
  labels:
    argocd.argoproj.io/secret-type: cluster
type: Opaque
stringData:
  name: destination-cluster
  server: <SERVER_URL_REDACTED>
  config: |
    {
      "awsAuthConfig": {
        "clusterName": "destination-cluster"
        // "roleARN": "arn:aws:iam::<AWS_ACCOUNT>:role/argo-role"  <-- not needed
      },
      "tlsClientConfig": {
        "caData": "<REDACTED>"
      }
    }
```

---

### Terraform examples

(we used `argocd_cluster` resource from [argocd provider](https://registry.terraform.io/providers/argoproj-labs/argocd/) instead of creating a secret)

**Destination cluster** – allow Argo CD role access and register the destination cluster to Argo

```hcl
resource "aws_eks_access_entry" "argocd" {
  cluster_name  = var.eks_cluster_name
  principal_arn = var.argocd_role_arn
  type          = "STANDARD"
}

resource "aws_eks_access_policy_association" "argocd" {
  cluster_name  = var.eks_cluster_name
  principal_arn = var.argocd_role_arn
  policy_arn    = "arn:aws:eks::aws:cluster-access-policy/AmazonEKSClusterAdminPolicy"

  access_scope {
    type = "cluster"
  }
}

resource "argocd_cluster" "eks_cluster" {
  server = var.eks_cluster_endpoint
  name   = var.argocd_cluster_name

  config {
    aws_auth_config {
      cluster_name = var.eks_cluster_name
      # for a single role - do not set role_arn here!
    }

    tls_client_config {
      ca_data = var.eks_cluster_ca_data
    }
  }
}
```

**Argo CD cluster – Pod Identity associations** - using eks-pod-identity community module, which creates the associations and the Argo role itself.

```hcl
module "argocd_pod_identity" {
  source  = "terraform-aws-modules/eks-pod-identity/aws"
  version = "2.6.0"

  name            = "${var.cluster_name}-argocd"
  use_name_prefix = false

  associations = {
    server = {
      cluster_name    = var.cluster_name
      namespace       = "argocd"
      service_account = "argocd-server"
    }
    controller = {
      cluster_name    = var.cluster_name
      namespace       = "argocd"
      service_account = "argocd-application-controller"
    }
  }
}
```
